### PR TITLE
Support for artifacts API

### PR DESCRIPTION
**Pull request by YannickJadoul**
This builds further on @JacekPliszka's work, but without the metaclasses to automatically add fields to a new class (cfr. #1556; which seems interesting, but unrelated to this actual functionality). See also #1373.

This still needs tests, but I cannot immediately find any artifacts belonging to this repository. Any idea what I can use to test downloading artifacts?

Things to review/consider:
- Following the status "302 Found" redirect, needed during the download of artifacts, is handled manually, since existing tests break when adding this to `Requester.__requestRaw` (see @JacekPliszka's f69c763, and my a7fd8f2). Is this something that should be preferably be added to `Requester.__requestRaw`? I can do an effort to fix those failing tests, then.
- I had to refactor the methods of `Requester` quite a bit, in order to be able to download a binary file without trying to decode it to a unicode string. While this seems like a lot of changes, I think they should be reasonably focused. See 99bbfbe and 78e7ae3.
- https://docs.github.com/en/free-pro-team@latest/rest/reference/actions#artifacts mentions that setting "`accept: application/vnd.github.v3+json` is recommended", yet this is not done for querying JSON in general. Should I do this for the new methods, or is this a separate PR to get this done for everything?

### CODE REVIEW

File: github/Artifact.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+            )
+
+        return data
+
+    def delete(self):
+        """
+        :calls: `DELETE /repos/:owner/:repo/actions/artifacts/:artifact_id <https://developer.github.com/v3/actions/artifacts/#delete-an-artifact>`_
+        :rtype: None
+        """
+        headers, data = self._requester.requestJsonAndCheck("DELETE", self.url)
```
Comment (s-t-e-v-e-n-k) [inline github/Artifact.py:137]: requestJson() and also check the returned status code is 204, like other examples in the code base.

File: github/Artifact.pyi [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+from typing import Any, Dict, Union
+
+from github.GithubObject import CompletableGithubObject
+
+class Artifact(CompletableGithubObject):
+    def __repr__(self) -> str: ...
+    def _initAttributes(self) -> None: ...
+    def _useAttributes(self, attributes: Dict[str, Any]) -> None: ...
+    def download(self, archive_format: str) -> bytes: ...
+    def delete(self) -> None: ...
```
Comment (s-t-e-v-e-n-k) [inline github/Artifact.pyi:11]: The above means this turns to bool.

File: github/WorkflowRun.py [hunk 3]
```diff
@@ -92,6 +95,14 @@ def conclusion(self):
         self._completeIfNotSet(self._conclusion)
         return self._conclusion.value
 
+    @property
+    def workflow_id(self):
```
Comment (s-t-e-v-e-n-k) [inline github/WorkflowRun.py:99]: This is unrelated to the Artifacts API, I'd prefer it was split out to its own tiny PR.

**Review by s-t-e-v-e-n-k — CHANGES_REQUESTED** [review]
This is looking like a great start, however the decode changes are incredibly far reaching, more so than I would have thought, you can't use __requestRaw? I have some other concerns inline.

File: github/WorkflowRun.py (hunk 3 above)
Comment (YannickJadoul) [reply github/WorkflowRun.py:99]: Good point. Done: #1737

File: github/Artifact.py (hunk 1 above)
Comment (YannickJadoul) [reply github/Artifact.py:137]: OK, sure. FYI, most of the things I found when I looked for `DELETE` were of this form, though? E.g.:

https://github.com/PyGithub/PyGithub/blob/1bb18ab52e9a5997774c9aed4e1521fe382e5b21/github/PullRequestComment.py#L172-L177

So that's probably why I ended up with this.

**Comment by YannickJadoul** [comment]
> This is looking like a great start, however the decode changes are incredibly far reaching, more so than I would have thought, you can't use __requestRaw?

I've tried to clean up a little bit further, yes. However, the main problem is that accessing the `requests.Response.text` attribute (inside `RequestsResonse`, see [here](https://github.com/PyGithub/PyGithub/blob/1bb18ab52e9a5997774c9aed4e1521fe382e5b21/github/Requester.py#L75)) tries to decode to Unicode (see [here in the Requests docs](https://requests.readthedocs.io/en/latest/api/#requests.Response.text)) and fails in the case of a zip file. To get `bytes`, we need to access [`requests.Response.content`](https://requests.readthedocs.io/en/master/api/#requests.Response.content). So we cannot directly use `__requestRaw` as it is now.

File: github/Artifact.py (hunk 1 above)
Comment (YannickJadoul) [reply github/Artifact.py]: I have changed it, at any rate; see 8ec83c32fa0be91d539a33803b0bddef99528f6d

File: github/Artifact.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+#                                                                              #
+# You should have received a copy of the GNU Lesser General Public License     #
+# along with PyGithub. If not, see <http://www.gnu.org/licenses/>.             #
+#                                                                              #
+################################################################################
+
+import github.GithubObject
+
+
+class Artifact(github.GithubObject.CompletableGithubObject):
```
Comment (yixinguo2) [inline github/Artifact.py:29]: Hi, before seeing your PR, I also did something similar and [created a PR](https://github.com/PyGithub/PyGithub/pull/1985). Please take a look and see if there are any useful parts that you could incorporate like some of the tests...

**Comment by bauergeorg** [comment]
Could you reopen this pull request?
